### PR TITLE
WIP: Pointers in cinterop

### PR DESCRIPTION
**Pull request by ilmat192**

### CODE REVIEW

File: Interop/Indexer/src/main/kotlin/org/jetbrains/kotlin/native/interop/indexer/MacroConstants.kt [hunk 1]
```diff
@@ -110,7 +114,7 @@ private fun reparseWithCodeSnippet(library: NativeLibrary,
         val codeSnippetLines = when (library.language) {
             // Note: __auto_type is a GNU extension which is supported by clang.
             Language.C, Language.OBJECTIVE_C -> listOf(
-                    "const __auto_type KNI_INDEXER_VARIABLE = $name;"
+                    "void kni_indexer_function() { __auto_type KNI_INDEXER_VARIABLE = $name; }"
```
Comment (SvyatoslavScherbina) [inline Interop/Indexer/src/main/kotlin/org/jetbrains/kotlin/native/interop/indexer/MacroConstants.kt]: Is it required to define a wrapping function?

File: Interop/Indexer/src/main/kotlin/org/jetbrains/kotlin/native/interop/indexer/MacroConstants.kt [hunk 2]
```diff
@@ -127,34 +131,41 @@ private fun processCodeSnippet(
...
-
+                evalResultOrNull = clang_Cursor_Evaluate(cursor)
+                state = VisitorState.EXPECT_VARIABLE_VALUE
                 CXChildVisitResult.CXChildVisit_Recurse
             }
 
             state == VisitorState.EXPECT_VARIABLE_VALUE && clang_isExpression(kind) != 0 -> {
                 typeOrNull = typeConverter(clang_getCursorType(cursor))
-                state = VisitorState.EXPECT_END
+                state = if (typeOrNull != null) {
```
Comment (SvyatoslavScherbina) [inline Interop/Indexer/src/main/kotlin/org/jetbrains/kotlin/native/interop/indexer/MacroConstants.kt]: `typeConverter` doesn't produce `null` values, AFAIK.

File: backend.native/tests/interop/basics/cmacros.def [hunk 3]
```diff
@@ -10,5 +24,19 @@
 #define ONE_POINT_ZERO 1.0
 #define ONE_POINT_FIVE 1.5f
 
+#define NULL_PTR (void*)0
```
Comment (SvyatoslavScherbina) [inline backend.native/tests/interop/basics/cmacros.def]: It makes sense to add brackets to this and following definitions to make them less ambiguous, i.e.
`#define NULL_PTR ((void*)0)`

File: Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/jvm/StubGenerator.kt [hunk 4]
```diff
@@ -848,6 +848,16 @@ class StubGenerator(
             }
         }
 
+        nativeIndex.wrappedMacros.forEach {
+            try {
+                stubs.add(
+                        GlobalVariableStub(GlobalDecl(it.name, it.type, isConst = true), this)
```
Comment (SvyatoslavScherbina) [inline Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/jvm/StubGenerator.kt]: This is ok, but `GlobalVariableStub` has special handling for struct- and array-typed C globals, and may generate incorrect stubs for such macros.
I suppose `GlobalVariableStub` could distinguish C macro and C global cases, and don't do special handling for the first one.

File: Interop/Indexer/src/main/kotlin/org/jetbrains/kotlin/native/interop/indexer/MacroConstants.kt (hunk 1 above)
Comment (ilmat192) [reply Interop/Indexer/src/main/kotlin/org/jetbrains/kotlin/native/interop/indexer/MacroConstants.kt]: Looks like it's required. Without the wrapping function the variable is global and we cannot use a function to initialize it. So we cannot expand macros like `#define FOO foo()`.

File: backend.native/tests/interop/basics/cmacros.def (hunk 3 above)
Comment (ilmat192) [reply backend.native/tests/interop/basics/cmacros.def]: Fixed.

File: Interop/Indexer/src/main/kotlin/org/jetbrains/kotlin/native/interop/indexer/MacroConstants.kt (hunk 1 above)
Comment (SvyatoslavScherbina) [reply Interop/Indexer/src/main/kotlin/org/jetbrains/kotlin/native/interop/indexer/MacroConstants.kt]: Btw, is `errno` supported now? On macOS it is declared like
```
#define errno (*__error())
```

Comment (ilmat192) [reply Interop/Indexer/src/main/kotlin/org/jetbrains/kotlin/native/interop/indexer/MacroConstants.kt]: Yes, it's supported now.

File: Interop/Indexer/src/main/kotlin/org/jetbrains/kotlin/native/interop/indexer/MacroConstants.kt (hunk 2 above)
Comment (ilmat192) [reply Interop/Indexer/src/main/kotlin/org/jetbrains/kotlin/native/interop/indexer/MacroConstants.kt]: Fixed

**Pull request merged**